Add thread sanitizer to test

NobodyXu opened the pull request:
This PR adds thread sanitizer to github workflow CI.

Signed-off-by: Jiahao XU <Jiahao_XU@outlook.com>

vorner commented on .github/workflows/test.yaml:
```diff
@@ -213,3 +213,29 @@ jobs:
+    - name: Run thread sanitizer
+      run: |
+        export RUSTDOCFLAGS='-Zsanitizer=thread'
+        export RUSTFLAGS='-Zsanitizer=thread'
```
Please use the `env` section for environment variables instead of shell exports.

NobodyXu replied:
Fixed.

vorner commented on .github/workflows/test.yaml:
```diff
@@ -213,3 +213,29 @@ jobs:
+      run: |
+        export RUSTDOCFLAGS='-Zsanitizer=thread'
+        export RUSTFLAGS='-Zsanitizer=thread'
+        for _ in $(seq 1 10); do cargo +nightly test -Z build-std --target --target $(uname -m)-unknown-linux-gnu; done
```
What is the for cycle for in here? If it really serves some particular purpose, a comment explaining why such weirdness is necessary would be nice.

NobodyXu replied:
Thread sanitizer isn't guaranteed to catch all data races, it can only catch it if the data race happens when running the program.

Running the tests multiple times increase the chances that data races are found by thread sanitizer.

NobodyXu replied:
I've added comment to explain the loop used here.

vorner replied:
Well, that's true for most of the tests in around this crate. Mostly, that's why the stress tests do multiple iterations internally.

But I guess it's OK.

vorner reviewed (changes requested):
Also, as mentioned in the linked ticket… I'm hesitant to merge it right now until we know if the reported failure is a false positive or not and until there's a solution. But once once that is solved, I'm happy to have it.

vorner commented:
Nice that the MacOS one works. But the UBuntu one would keep the CI red all the time, so it's not particularly useful.

Do you agree if I cherry-pick that one and leave the rest? Or do you want to create a cleaned-up branch with that one yourself?

NobodyXu commented:
> Nice that the MacOS one works. But the UBuntu one would keep the CI red all the time, so it's not particularly useful.
> 
> Do you agree if I cherry-pick that one and leave the rest? Or do you want to create a cleaned-up branch with that one yourself?

I will remove the ubuntu thread sanitizer test since it is likely to be just false positive and really isn't helpful.

NobodyXu commented:
@vorner I've removed the Ubuntu thread sanitizer and rebased the branch again vorner:master.

All the [workflow](https://github.com/NobodyXu/arc-swap/actions) succeeded.
